versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/indexes/StoreKey.java

```diff
@@ -39,24 +41,54 @@ public final class StoreKey implements Comparable<StoreKey> {
   public static final int MAX_LENGTH = 500;
 
   /** Contains the {@code char 0} separated key representation. */
-  private final String key;
+  private final byte[] key;
 
-  private StoreKey(String key) {
+  private final boolean prefixEnd;
+
+  private StoreKey(byte[] key, boolean prefixEnd) {
     this.key = key;
+    this.prefixEnd = prefixEnd;
+  }
+
+  public byte[] rawBytes() {
+    return Arrays.copyOf(key, key.length);
+  }
+
+  public ByteString rawByteString() {
+    return ByteString.copyFrom(key);
   }
 
   public String rawString() {
-    return key;
+    return new String(key, UTF_8);
+  }
+
+  public int rawLength() {
+    return key.length;
   }
 
   @Override
   public int compareTo(StoreKey that) {
-    return key.compareTo(that.key);
+    int diff = ByteBuffer.wrap(key).mismatch(ByteBuffer.wrap(that.key));
+    if (diff == -1) {
+      return 0;
+    }
+    if (diff >= key.length) {
+      if (prefixEnd && diff == key.length) {
+        return 1;
+      }
+      return -1;
+    }
+    if (diff >= that.key.length) {
+      return 1;
+    }
```

Comment:
This makes comparisons non-commutative. It will bite us sooner or later, I think.
```
  @Test
  void comparePrefix() {
    StoreKey k1 = StoreKey.keyFromStringPrefixEnd("aaa");
    StoreKey k2 = StoreKey.keyFromStringPrefixEnd("aaaBBB");
    soft.assertThat(k1).isGreaterThan(k2);
    soft.assertThat(k2).isLessThan(k1); // this assert fails
  }
```

Reply:
I think it might be preferable to actually encode the "prefix" flag in the key's byte form (to ensure consistency in case it ever gets into storage).